### ISSUE TITLE
Support for AzureAD to connect to Storage Table APi's

### ISSUE THREAD

**Issue by marcoboffi** [hashicorp/terraform-provider-azurerm#15083]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Currently _storage_use_azuread_ flag in provider configuration works only with Blob and Queue API's.  
Authorization with Azure AD for Table is [now available](https://docs.microsoft.com/en-us/azure/storage/tables/authorize-access-azure-active-directory) in preview (in ALL region).  
Can you add support also for Storage Table API's ?

### New or Affected Resource(s)

* azurerm_storage_table, azurerm_storage_table_entity

### Potential Terraform Configuration

```hcl
# Copy-paste your Terraform configurations here - for large Terraform configs,
# please use a service like Dropbox and share a link to the ZIP file. For
# security, you can also encrypt the files using our GPG public key.
```

### References

* #0000

**Comment by magodo** [comment]
@marcoboffi Thank you for submitting this :+1: During the development on this feature, I found that the API (at least for 2020-08-04) only supports for the table operations, but not the ACL operations of the table, which is necessary given current implementation of the provider.

See: https://github.com/Azure/azure-rest-api-specs/issues/17485

**Comment by beau-witter** [comment]
Based on the table on [this page](https://learn.microsoft.com/en-us/azure/storage/common/authorize-data-access?toc=%2Fazure%2Fstorage%2Fblobs%2Ftoc.json&bc=%2Fazure%2Fstorage%2Fblobs%2Fbreadcrumb%2Ftoc.json#understand-authorization-for-data-operations) it appears that at least Blob, Table, and Queues can be managed via Azure AD now. It seems there is at least _a_ way to do it for the File Shares as well (at least as SMB), so maybe that could be added as well?

**Comment by commakoerschgen** [comment]
I can confirm from my own experience that manipulation of table data is possible with Azure AD authentication.

**Comment by marcoboffi** [comment]
I don't understand if implementation is blocked by [this](https://github.com/Azure/azure-rest-api-specs/issues/17485#issuecomment-1090995403) or not.

**Comment by manicminer** [comment]
The Storage Tables API now supports AAD authentication _to some extent_ and this will be partially enabled by #24798. Unfortunately, setting and/or retrieving ACLS for Storage Tables does not support AAD authentication at this time (see https://github.com/Azure/azure-rest-api-specs/issues/17485#issuecomment-1090995403 as linked by @marcoboffi), this means that whilst AAD auth will be enabled for the `azurerm_storage_table_entity` resource & data sources, it won't yet be fully enabled for the `azurerm_storage_table` resource.

AAD auth support also won't extend to Files at this time, as contrary to official documentation, this API does not properly support AAD authentication as yet.

**Comment by jan-hudec** [comment]
> Unfortunately, setting and/or retrieving ACLS for Storage Tables does not support AAD authentication at this time

My understanding is that:

1. that's the same situation as for Blobs, and
2. ACLs are only applicable to key access, so if you want AAD (Entra) authentication because you want to turn the key authentication off, you won't be using the ACLs anyway (and maybe using Entra role assignments intead).

**Comment by jan-hudec** [comment]
@manicminer, the main reason to want Entra/AAD auth is to disable the key authentication altogether, in which case the reading of ACLs is going to fail. So it should be considered not a failure, but absence of any ACLs. (The proof of concept where I had the requirement has long since been scratched, but others will have similar requirement for security reasons.)

**Comment by tombuildsstuff** [comment]
@jan-hudec it's the migration paths which are going to complicate matters, since you can have both Entra and Shared Key access enabled either individually or simultaneously.

In addition we'd need to validate what happens to any ACLs when using both Shared Key and Entra auth - and when moving from Shared Key -> Entra -> Shared Key, it's not _necessarily_ sufficient to just skip the check, since we'd want to be able to detect any changes here

**Comment by jkroepke** [comment]
To unblock this kind of issue, would it be possible to move ACLs to an dedicated resource? Then, azurerm_storage_table could be used with AzureAD and users which is using ACL can use shared key together with the dedicated terraform resource.

**Comment by ghidalgo3** [comment]
Is this issue being worked on? Using AAD credentials for all APIs is a security requirement.

**Comment by marcoboffi** [comment]
> > I don't understand if implementation is blocked by [this](https://github.com/Azure/azure-rest-api-specs/issues/17485#issuecomment-1090995403) or not.
> 
> Unfortunately, it is.
> 
> The Storage Tables API now supports AAD authentication _to some extent_ and this will be partially enabled by [#24798](https://github.com/hashicorp/terraform-provider-azurerm/pull/24798). Unfortunately, setting and/or retrieving ACLS for Storage Tables does not support AAD authentication at this time - this means that whilst AAD auth will be enabled for the `azurerm_storage_table_entity` resource & data sources, it won't yet be fully enabled for the `azurerm_storage_table` resource.
> 
> AAD auth support also won't extend to Files at this time, as this API does not properly support AAD authentication as yet.
> 
> See [official docs](https://learn.microsoft.com/en-us/rest/api/storageservices/authorize-with-azure-active-directory) for a breakdown of supported auth methods for different operations.

https://azure.microsoft.com/it-it/updates/?id=496287

**Comment by khharper** [comment]
@xgithubtriage this behavior also impacts Azure Files.
Our team can’t disable Allow storage account key access because the azurerm_storage_share resource still performs management operations with Shared Key. We need AAD (Microsoft Entra) auth for File Share management (create/update/delete) in the provider.
Reference: the registry notes for azurerm_storage_share/data.azurerm_storage_share call out Shared Key behavior, which prevents AAD‑only scenarios, and there’s prior discussion in #17341 about KeyBasedAuthenticationNotPermitted when keys are disabled.

I’ll open a separate issue dedicated to File Shares and link it here for tracking. Based on this https://learn.microsoft.com/en-us/rest/api/storageservices/authorize-with-azure-active-directory my understanding is Entra based management can be supported now

**Comment by marcoboffi** [comment]
I see MR is blocked from several time.
Has the feature been merged via another pull request, or is it still missing?

**Comment by magodo** [comment]
Related to: #30815.

**Comment by s4parke** [comment]
> [@jan-hudec](https://github.com/jan-hudec) it's the migration paths which are going to complicate matters, since you can have both Entra and Shared Key access enabled either individually or simultaneously.
> 
> In addition we'd need to validate what happens to any ACLs when using both Shared Key and Entra auth - and when moving from Shared Key -> Entra -> Shared Key, it's not _necessarily_ sufficient to just skip the check, since we'd want to be able to detect any changes here

There is no need to validate what happens-- it is [explained in the MSFT documentation](https://learn.microsoft.com/en-us/azure/storage/files/authorize-data-operations-portal). 

When shared keys are used for authentication, the caller gets "super-user" permission regardless of what other authorization scopes are configured on the resource.
When Entra ID with RBAC role assignments are used, they take precedence over any ACLs that are present.

Shared access keys are the number one attack vector for cloud data exposure. A contributing factor is outdated infrastructure. I came here and found the root cause.
